bpf/lib/lb.h

```diff
@@ -689,8 +689,54 @@ struct lb6_service *lb6_lookup_backend_slot(struct __ctx_buff *ctx __maybe_unuse
 	return NULL;
 }
 
+#ifdef PER_SVC_LB
+static __always_inline __u32
+lb6_select_backend_id_random(struct __ctx_buff *ctx,
+			     struct lb6_key *key,
+			     const struct ipv6_ct_tuple *tuple __maybe_unused,
+			     const struct lb6_service *svc)
+{
+	__u16 slot = (get_prandom_u32() % svc->count) + 1;
+	struct lb6_service *be = lb6_lookup_backend_slot(ctx, key, slot);
+
+	return be ? be->backend_id : 0;
+}
+
+static __always_inline __u32
+lb6_select_backend_id_maglev(struct __ctx_buff *ctx __maybe_unused,
+			     struct lb6_key *key __maybe_unused,
+			     const struct ipv6_ct_tuple *tuple,
+			     const struct lb6_service *svc)
+{
+	__u32 zero = 0, index = svc->rev_nat_index;
+	__u32 *backend_ids;
+	void *maglev_lut;
+
+	maglev_lut = map_lookup_elem(&LB6_MAGLEV_MAP_OUTER, &index);
+	if (unlikely(!maglev_lut))
+		return 0;
+
+	backend_ids = map_lookup_elem(maglev_lut, &zero);
+	if (unlikely(!backend_ids))
+		return 0;
+
+	index = hash_from_tuple_v6(tuple) % LB_MAGLEV_LUT_SIZE;
+	return map_array_get_32(backend_ids, index, (LB_MAGLEV_LUT_SIZE - 1) << 2);
+}
+
+static __always_inline __u32
+lb6_select_backend_id(struct __ctx_buff *ctx,
+		      struct lb6_key *key,
+		      const struct ipv6_ct_tuple *tuple __maybe_unused,
+		      const struct lb6_service *svc)
+{
+	if (svc->lb_alg == LB_SELECTION_MAGLEV)
+		return lb6_select_backend_id_maglev(ctx, key, tuple, svc);
+
+	return lb6_select_backend_id_random(ctx, key, tuple, svc);
+}
+#elif LB_SELECTION == LB_SELECTION_RANDOM
 /* Backend slot 0 is always reserved for the service frontend. */
-#if LB_SELECTION == LB_SELECTION_RANDOM
 static __always_inline __u32
 lb6_select_backend_id(struct __ctx_buff *ctx,
```

Comment:
We could later consolidate this, e.g. `#define lb6_select_backend_id lb6_select_backend_id_random` or such. I think follow-up is fine.